File: 04--scratchcards/a.hs
#!/usr/bin/env runhaskell
-- Usage:
--   ./a.hs PATH_TO_INPUT_FILE

import Data.Function ((&))
import System.Environment (getArgs)
import qualified Data.Set as Set


main = do
  [inputPath] <- getArgs
  text <- readFile inputPath
  print $ solve text


solve text =
  text
  & lines
  & map solveLine
  & sum


solveLine card =
  card

  -- Take only after the ':'
  & dropWhile (/= ':')
  & drop 1

  -- Split on '|' and parse ints
  & map (\c -> if c == '|' then '\n' else c)
  & lines
  & map readIntsToSet

  -- Count matches
  & \[winners, chosen] -> Set.intersection winners chosen
  & Set.size

  -- Calculate points for this card
  & \matches ->
      if matches == 0
      then 0
      else 2 ^ (matches - 1)


-- e.g. readIntsToSet "1 2 3" == Set.fromList [1, 2, 3]
readIntsToSet s =
  s
  & words
  & map readInt
  & Set.fromList


readInt = read :: String -> Int

File: 04--scratchcards/a.py
#!/usr/bin/env python3
'''
Usage:
    ./a.py PATH_TO_INPUT_FILE
    ./b.py PATH_TO_INPUT_FILE
'''

from sscanf import sscanf

import sys
from dataclasses import dataclass


def main():
    cards = []
    for line in open(sys.argv[1]).read().splitlines():
        _, winners, chosen = sscanf(line, 'Card %s: %s | %s')
        winners = {int(n) for n in winners.strip().split()}
        chosen = {int(n) for n in chosen.strip().split()}
        cards.append(Card(winners, chosen))

    answer = 0
    for card in cards:
        matches = len(card.winners & card.chosen)
        if matches:
            points = 2 ** (matches - 1)
        else:
            points = 0
        answer += points

    print(answer)


@dataclass
class Card:
    winners: set[int]
    chosen: set[int]


if __name__ == '__main__':
    main()
